presto-main/src/main/java/com/facebook/presto/execution/SharedBufferMemoryManager.java

```diff
@@ -30,12 +40,36 @@ public SharedBufferMemoryManager(long maxBufferedBytes, SystemMemoryUsageListene
         checkArgument(maxBufferedBytes > 0, "maxBufferedBytes must be > 0");
         this.maxBufferedBytes = maxBufferedBytes;
         this.systemMemoryUsageListener = requireNonNull(systemMemoryUsageListener, "systemMemoryUsageListener is null");
+
+        notFull = SettableFuture.create();
+        notFull.set(null);
     }
 
     public void updateMemoryUsage(long bytesAdded)
     {
         systemMemoryUsageListener.updateSystemMemoryUsage(bytesAdded);
         bufferedBytes.addAndGet(bytesAdded);
+        if (!isFull()) {
+            synchronized (this) {
+                if (!isFull() && !notFull.isDone()) {
+                    notFull.set(null);
+                }
+            }
+        }
+    }
+
+    public synchronized ListenableFuture<?> getNotFullFuture()
+    {
+        if (isFull() && notFull.isDone()) {
+            notFull = SettableFuture.create();
+        }
+        return notFull;
+    }
+
+    public void setNoBlockOnFull()
```

Comment:
When the buffer is destroyed (say due to a limit), we need free the writers (and make sure no one else blocks).  This is the least error prone method I could think of to accomplish this.